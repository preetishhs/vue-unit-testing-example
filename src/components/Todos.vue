<template>
  <section>
    <h1>TO DO LIST</h1>
    <div class="todoapp">
      <header class="header">
        <input
          class="new-todo"
          autofocus
          autocomplete="off"
          placeholder="Add an item"
          @keyup.enter="addTodo"
        />
      </header>
      <section class="main" v-show="todoList.length">
        <ul class="todo-list">
          <li
            v-for="(item, index) in todoList"
            class="todo"
            :key="item.id"
            :class="{ completed: item.completed }"
            @click="toggleState(item.id)"
          >
            <div class="view">
              <label> {{ item.title }}</label>
              <button class="destroy" @click.stop="removeTodo(index)"></button>
            </div>
          </li>
        </ul>
      </section>
      <footer class="footer" v-show="todoList.length">
        <div class="total">Items: {{ todoList.length }}</div>
        <button class="async-add" @click="addItemFromWeb">Async Add</button>
        <div class="completed-list">Completed: {{ completedList.length }}</div>
      </footer>
    </div>
  </section>
</template>

<script>
import { mapGetters, mapMutations, mapActions, mapState } from 'vuex'

export default {
  name: 'todos',
  data() {
    return {}
  },

  computed: {
    ...mapGetters(['completedList']),
    ...mapState({
      todoList: state => state.todos
    })
  },

  methods: {
    ...mapMutations(['addItem', 'toggleCompletion', 'removeItem']),

    ...mapActions(['addItemFromWeb']),

    toggleState(index) {
      return this.toggleCompletion(index)
    },
    addTodo(event) {
      if (!event.target.value.trim()) {
        return
      }
      this.addItem(event.target.value)
      event.target.value = ''
    },
    removeTodo: function(index) {
      this.removeItem(index)
    }
  }
}
</script>

<style lang="scss" scoped>
button.destroy {
  margin: 0;
  padding: 0;
  border: 0;
  background: none;
  font-size: 100%;
  vertical-align: baseline;
  font-family: inherit;
  font-weight: inherit;
  color: inherit;
  -webkit-appearance: none;
  appearance: none;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

:focus {
  outline: 0;
}

.hidden {
  display: none;
}

.todoapp {
  width: 90%;
  background: #fff;
  margin: 80px auto 40px auto;
  position: relative;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.2), 0 25px 50px 0 rgba(0, 0, 0, 0.1);
}

.todoapp input::-webkit-input-placeholder {
  font-style: italic;
  font-weight: 300;
  color: #a5a5a5;
}

.todoapp input::-moz-placeholder {
  font-style: italic;
  font-weight: 300;
  color: #a5a5a5;
}

.todoapp input::input-placeholder {
  font-style: italic;
  font-weight: 300;
  color: #a5a5a5;
}

h1 {
  width: 100%;
  font-size: 60px;
  font-weight: 100;
  text-align: center;
  margin-top: 80px;
  color: #fff;
}

.new-todo,
.edit {
  position: relative;
  margin: 0;
  width: 100%;
  font-size: 24px;
  font-family: inherit;
  font-weight: inherit;
  line-height: 1.4em;
  border: 0;
  color: inherit;
  padding: 6px;
  border: 1px solid #999;
  box-shadow: inset 0 -1px 5px 0 rgba(0, 0, 0, 0.2);
  box-sizing: border-box;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.new-todo {
  padding: 16px 16px 16px 16px;
  border: none;
  background: rgba(0, 0, 0, 0.003);
  box-shadow: inset 0 -2px 1px rgba(0, 0, 0, 0.03);
}

.main {
  position: relative;
  z-index: 2;
  border-top: 1px solid #e6e6e6;
}

.todo-list {
  margin: 0;
  padding: 0;
  list-style: none;
}

.todo-list li {
  position: relative;
  font-size: 24px;
  border-bottom: 1px solid #ededed;
}

.view label {
  cursor: pointer;
}

.todo-list li:last-child {
  border-bottom: none;
}

.todo-list li label {
  word-break: break-all;
  padding: 15px 60px 15px 15px;
  display: block;
  line-height: 1.2;
  transition: color 0.4s;
}

.todo-list li.completed label {
  color: #a5a5a5;
  text-decoration: line-through;
}

.todo-list li .destroy {
  display: none;
  position: absolute;
  top: 0;
  right: 10px;
  bottom: 0;
  width: 40px;
  height: 40px;
  margin: auto 0;
  font-size: 30px;
  color: #53f5ac;
  margin-bottom: 11px;
  transition: color 0.2s ease-out;
}

.todo-list li .destroy:hover {
  color: #af5b5e;
}

.todo-list li .destroy:after {
  content: 'Ã—';
}

.todo-list li:hover {
  background: #d7ffed;
}

.todo-list li:hover .destroy {
  display: block;
}

.todo-list li .edit {
  display: none;
}

.todo-list li.editing:last-child {
  margin-bottom: -1px;
}

footer {
  height: 40px;
  display: flex;
  border-top: 1px solid #ededed;
  background: #ededed;
  align-items: center;
  justify-content: space-around;
  font-size: 16px;
}
</style>
