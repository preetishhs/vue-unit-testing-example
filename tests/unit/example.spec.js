import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import Todos from '@/components/Todos.vue'

const localVue = createLocalVue()

localVue.use(Vuex)

let store, state, getters, mutations, actions

beforeEach(() => {
  state = {
    todos: []
  }
  getters = {
    completedList: () => []
  }
  mutations = {
    addItem: jest.fn(),
    toggleCompletion: jest.fn(),
    removeItem: jest.fn()
  }
  actions = {
    addItemFromWeb: jest.fn()
  }
  store = new Vuex.Store({
    state,
    getters,
    mutations,
    actions
  })
})

describe('todo UI checks', () => {
  test('Check if input box is present', () => {
    const wrapper = shallowMount(Todos, { store, localVue })
    expect(wrapper.contains('input.new-todo')).toBe(true)
  })

  test('check if one item is present', () => {
    store.state.todos = [{ id: 1, title: 'one item', completed: false }]
    const wrapper = shallowMount(Todos, { store, localVue })
    expect(wrapper.contains('li.todo')).toBe(true)
  })

  test('check if an item has been crossed', () => {
    store.state.todos = [{ id: 1, title: 'one item', completed: true }]
    const wrapper = shallowMount(Todos, { store, localVue })
    expect(wrapper.contains('li.completed')).toBe(true)
  })

  test('check if remove button is present', () => {
    store.state.todos = [{ id: 1, title: 'one item', completed: true }]
    const wrapper = shallowMount(Todos, { store, localVue })
    expect(wrapper.contains('button.destroy')).toBe(true)
  })
})

describe('vuex component level checks', () => {
  test('check if toggleCompletion mutation was called on click', () => {
    store.state.todos = [{ id: 1, title: 'one item', completed: true }]
    const wrapper = shallowMount(Todos, { store, localVue })
    wrapper.find('li.todo').trigger('click')
    expect(mutations.toggleCompletion).toHaveBeenCalled()
  })

  test('check if addItem was not called for empty value on keyboard enter', () => {
    const wrapper = shallowMount(Todos, { store, localVue })
    wrapper.find('input.new-todo').trigger('click')
    wrapper.find('input.new-todo').trigger('keyup.enter')
    expect(mutations.addItem).not.toHaveBeenCalled()
  })

  test('check if addItem was called for non empty value on keyboard enter', () => {
    const wrapper = shallowMount(Todos, { store, localVue })
    wrapper.find('input.new-todo').setValue('test')
    wrapper.find('input.new-todo').trigger('keyup.enter')
    expect(mutations.addItem).toHaveBeenCalled()
  })

  test('check if removeItem is called on clicking x button', () => {
    store.state.todos = [{ id: 1, title: 'one item', completed: false }]
    const wrapper = shallowMount(Todos, { store, localVue })
    wrapper.find('button.destroy').trigger('click')
    expect(mutations.removeItem).toHaveBeenCalled()
  })

  test('check if addItemFromWeb action was called on button click', () => {
    // for that button to be visible we need one entry
    store.state.todos = [{ id: 1, title: 'one item', completed: false }]
    const wrapper = shallowMount(Todos, { store, localVue })
    wrapper.find('button.async-add').trigger('click')
    expect(actions.addItemFromWeb).toHaveBeenCalled()
  })
})

describe('Todo snapshot checks', () => {
  test('check if input box renders correctly', () => {
    const wrapper = shallowMount(Todos, { store, localVue })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
