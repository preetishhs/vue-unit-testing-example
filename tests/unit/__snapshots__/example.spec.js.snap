// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo snapshot checks check if input box renders correctly 1`] = `
<section>
  <h1>TO DO LIST</h1>
  <div class="todoapp">
    <header class="header"><input autofocus="autofocus" autocomplete="off" placeholder="Add an item" class="new-todo"></header>
    <section class="main" style="display: none;">
      <ul class="todo-list"></ul>
    </section>
    <footer class="footer" style="display: none;">
      <div class="total">Items: 0</div> <button class="async-add">Async Add</button>
      <div class="completed-list">Completed: 0</div>
    </footer>
  </div>
</section>
`;
